### DISCUSSION TITLE
Templates with multiple endpoints

### DISCUSSION THREAD

**Discussion by Paril** [mikke89/RmlUi#560] [Ideas]
Hey there.

I'm finally getting around to starting our first proper implementation of RMLUI in a project. One of the first things I've run into is a fairly simple concept - the template system is really nice, but also really limiting in that it can only replace one element (the "content" id); this means for something like a basic windowing system, you're forced to take many extra steps just to, for instance, set the title of a templated window.

Conceptually, I wonder if templates could be expanded to support multiple endpoints, so that a template like this:
```
<body class="window">
	<div id="window">
		<div id="title"></div>
		<div id="content"></div>
	</div>
</body>
```

could have both the title and content element filled at load time. I think the most convenient and obvious way to approach this would be to just have templates replace matching IDs, so for the template above, you could fill both in like this:

```
	<body template="window">
		<div id="title">$m_main_menu</div>
		<div id="content">
			<div class="item">$m_new_game</div>
			<div class="item">$m_load_game</div>
			<div class="item">$m_options</div>
			<div class="item">$m_credits</div>
			<div class="item">$m_quit</div>
		</div>
	</body>
```

Of course this is just a very basic minimal sample, it doesn't make sense at this scope since you might as well just replace the whole content anyways, but templates as they are seem very limiting to what they can offer which seems against the purpose of re-use if we have to do a bunch of boilerplate on the code side; I wonder if expanding the scope very slightly (in a backwards compatible way; obv this sample above isn't) would be a benefit.

**Comment by mikke89** [comment]
Thanks for the suggestion, this does seem like a handy little feature. Generally, I'd be happy to take a pull request for something like this.

There are some design decisions that should be discussed. As you mention, this wouldn't be backward compatible. Maybe in the document, instead of using `<div id="title">` we could use e.g. `<template for="title">`, or something like that, I'm sure the name could be improved. This could be confusing with the existing `<template>` element though. And, we should also add this feature to the latter if we add it to the `body template`.

**Reply by Paril** [reply]
Yeah I'll have to think a bit more about this one; I like the idea of smaller documents being used as reusable components, sort of like making custom elements but entirely in RML. Maybe they could be called components or something just to keep them away from the "template" nomenclature.
